dsc_lib/src/parser/functions.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use serde_json::Value;
+use tree_sitter::Node;
+
+use crate::DscError;
+use crate::parser::{
+    expressions::Expression,
+    FunctionDispatcher,
+};
+
+#[derive(Clone)]
+pub struct Function<'a> {
+    name: String,
+    args: Option<Vec<FunctionArg<'a>>>,
+    function_dispatcher: &'a FunctionDispatcher,
+}
+
+#[derive(Clone)]
+pub enum FunctionArg<'a> {
+    String(String),
+    Integer(i32),
+    Boolean(bool),
+    Expression(Expression<'a>),
+}
+
+#[derive(Debug, PartialEq)]
+pub enum FunctionResult {
+    String(String),
+    Object(Value),
+}
+
+impl<'a> Function<'a> {
+    /// Create a new `Function` instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `function_dispatcher` - The function dispatcher to use.
+    /// * `statement` - The statement that the function is part of.
+    /// * `function` - The function node.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the function node is not valid.
+    pub fn new(function_dispatcher: &'a FunctionDispatcher, statement: &str, function: &Node) -> Result<Self, DscError> {
+        let Some(function_name) = function.child_by_field_name("name") else {
+            return Err(DscError::Parser("Function name node not found".to_string()));
+        };
+        let function_args = function.child_by_field_name("args");
+        let args = convert_args_node(function_dispatcher, statement, &function_args)?;
+        Ok(Function{
+            function_dispatcher,
+            name: function_name.utf8_text(statement.as_bytes())?.to_string(),
+            args})
+    }
+
+    /// Invoke the function.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the function fails to execute.
+    pub fn invoke(&self) -> Result<FunctionResult, DscError> {
+        // if any args are expressions, we need to invoke those first
+        let mut resolved_args: Vec<FunctionArg> = vec![];
+        if let Some(args) = &self.args {
+            for arg in args {
+                match arg {
+                    FunctionArg::Expression(expression) => {
+                        let value = expression.invoke()?;
+                        resolved_args.push(FunctionArg::String(value));
+                    },
+                    _ => {
+                        resolved_args.push(arg.clone());
+                    }
+                }
+            }
+        }
+
+        self.function_dispatcher.invoke(&self.name, &resolved_args)
```

Comment:
same here, the `invoke` method should take a `& FunctionDispatcher` parameter, and call `invoke` on it here.